canonical/cloud-init
# Azure: Clean up and refactor report_diagnostic_event

## Proposed Commit Message

Azure helper: Refactor report_diagnostic_event and logging

## Additional Context

Across the cloud-init Azure datasource codebase, one of the `LOG` functions (`debug`, `warning`, and `error`) is always invoked whenever `report_diagnostic_event` is invoked. This changes `report_diagnostic_event` so that a logger function can be passed in, and then invoked within `report_diagnostic_event`.

## Test Steps
* Deploy an Azure VM (from any cloud-init enabled Ubuntu LTS image)
* Apply this PR's patch
* Reboot
* Check for a common message that is both (1) reported to KVP diagnostics and (2) logged to `cloud-init.log`. A good candidate is the `found datasource in` diagnostic message.
* Observe within `cloud-init.log` the logged diagnostic messages.
* Observe within `/var/lib/hyperv/.kvp_pool_1` (KVP diagnostics file) the logged diagnostic messages.
* See https://paste.ubuntu.com/p/TxpVCfckQc/

## Checklist:
